### PR TITLE
plugins: command-regexp functions are now private

### LINKED ISSUE #1887: tools: command-regexp functions shouldn't be public
The major rule-system rework in #1873 is a great excuse to take the `get_*_regexp()` and `get_*_pattern()` functions from `sopel.tools` and either shove them into someplace like `sopel.plugins.rules.utils` or encapsulate the pattern generation into the different rule object classes.

@Exirel let's come back to this after #1873 is finished.

**Full to-deprecate list (for 7.1):**
- [x] `compile_rule()` [#2027]
- [x] `get_command_regexp()` [#1944]
- [x] `get_command_pattern()` [#1944]
- [x] `get_nickname_command_regexp()` [#1944]
- [x] `get_nickname_command_pattern()` [#1944]
- [x] `get_action_command_regexp()` [#1944]
- [x] `get_action_command_pattern()` [#1944]

**Full to-delete list (for 8.0):**
- [x] `compile_rule()` [#2141]
- [x] `get_command_regexp()` [#2141]
- [x] `get_command_pattern()` [#2141]
- [x] `get_nickname_command_regexp()` [#2141]
- [x] `get_nickname_command_pattern()` [#2141]
- [x] `get_action_command_regexp()` [#2141]
- [x] `get_action_command_pattern()` [#2141]

### PR DESCRIPTION
**Pull request by Exirel**
### Description

Fix #1887 

* moved everything into `sopel.plugins.rules`, directly into each rule class
* removed/deprecated a bunch of stuff from `sopel.test_tools` (this entire module should probably be deprecated and removed at some point anyway)
* reworked imports in `sopel.test_tools`
* changed how the `example` decorator runs a test (this should be transformed further, probably in a future PR, for 7.1 or 8.x)
* removed obsolete tests and fixtures
* made sure the doc is still OK

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:1252]: I must have forgotten to mention it in that issue you were working from, but the new rule system is a perfect excuse to deprecate `tools.compile_rule()` too, and move its behavior into the `Rule` type.

**Review by dgw — CHANGES_REQUESTED** [review]
I have to ask: Are all of these changes _really_ related to the PR/commit title, `plugins: command-regexp functions are now private`? This is a rare occasion when I really think you're doing too much in one commit, unless all of the refactoring and such is truly necessary to do all at once. I'd say at least some of it probably isn't, and should be a separate commit or PR.

Two things about the overall approach here, though:

1. The existing functions in `tools` should be marked deprecated until removal in 8.0, instead of yanked out in a point release with no warning. But nothing says we have to keep their implementation; what would you say to building the implementation inside `Rule` and its subclasses such that those `tools` functions would just call a classmethod or act as an accessor for the appropriate class constant?
2. Can we put `tools.compile_rule` into `plugins.rules` too? Same thing about deprecation vs. outright removal in 7.1, but it has no reason to exist outside the rule system _especially_ if the rule system is its primary user.

File: sopel/tools/__init__.py
Comment (Exirel) [reply sopel/tools/__init__.py:352]: OK I'll rework that.

**Comment by Exirel** [comment]
> This is a rare occasion when I really think you're doing too much in one commit

I totally got carried away with that one, with a very bad case of tunnel vision. I'll see if I can separate that properly.

File: sopel/plugins/rules.py
Comment (Exirel) [reply sopel/plugins/rules.py:1252]: Hm. It requires a bit more work. I shall do that in another PR :tm:

File: sopel/tools/__init__.py
Comment (Exirel) [reply sopel/tools/__init__.py]: I get these functions back with a deprecation warning!

**Comment by Exirel** [comment]
> I'll see if I can separate that properly.

Just did! I had to rebase everything, so I took that opportunity to fix the merge conflict as well.

File: sopel/test_tools.py
Comment (dgw) [inline sopel/test_tools.py:263]: I think this is what you meant (but it's already been the case for a _long_ time, so we probably should've deprecated this sooner 😅)

```suggestion
@tools.deprecated('pytest now runs @plugin.example tests directly', '7.1', '8.0')
```

I want to mention `@module.example` too somehow, but can't think of a way that doesn't make the deprecation message extra long and cluttered. :/

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:64]: Why is this ported over to here? Nothing references the name `unicode` that I can see.

File: sopel/plugins/rules.py
Comment (dgw) [reply sopel/plugins/rules.py:1252]: I'll probably make an issue for this, because with no plans to release any 7.2 branch, our last chance to deprecate anything else we want to remove in 8.0 (like `tools.compile_rule`) is 7.1.

File: test/test_tools.py
Comment (dgw) [inline test/test_tools.py:8]: Won't we get a drop in coverage by removing these tests while the tested functions still exist? Coveralls reported -0.1% for the current patch. I think deleting these tests is premature.

I also think it's important to make sure the equivalent behavior is tested in `plugins.rules` and its various classes. Obviously `<RuleTypeClass>.get_rule_regex()` must be called in the course of setting up the tests in `test/plugins/test_plugins_rules.py`, so its behavior in each class is validated—but there's e.g. minimal validation of command rules' match groups (AFAICT, only groups 0 & 1).

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:226]: I'm gonna wordsmith a bit in this file. Thought about just making the suggestion once, but it's probably easier on you if I suggest it everywhere so you can just batch-apply the change if you like it. 😁

```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:244]:
```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:274]:
```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:293]:
```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:318]:
```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

File: sopel/tools/__init__.py
Comment (dgw) [inline sopel/tools/__init__.py:330]:
```suggestion
@deprecated('command regexp tools are now private', '7.1', '8.0')
```

**Review by dgw — COMMENTED** [review]
Overall not a whole lot to say; mostly repeated suggestions for wording that's used in a half-dozen spots.

I complained about removing tests for the deprecated `tools` functions, but there was no place to leave a line note about making them just call the corresponding `plugins.rules` implementations until removal (what I suggested before).

File: sopel/plugins/rules.py
Comment (Exirel) [reply sopel/plugins/rules.py:64]: Oh weird. I think an initial version used that. I shall remove that!

File: test/test_tools.py
Comment (Exirel) [reply test/test_tools.py:8]: I consider this:

* deprecated functions **must not** be modified once they are deprecated
* these functions have been tested before deprecation without issues

So I was a bit quick to delete these tests, I admit.

However, about the actual tests: https://github.com/sopel-irc/sopel/blob/b16cfcd8e1197e4e85727f00f56c69f01c890aac/test/plugins/test_plugins_rules.py#L1421-L1501

So it partially cover this, because only the commands are, not the nickname/action commands.

Comment (dgw) [reply test/test_tools.py:8]: Honestly, though, these "deprecated functions" aren't actually deprecated until this PR gets merged, so you are still allowed to change them. 😝 But my main concern is behavior drift, making sure the deprecated functions _and_ the new stuff in `plugins.rules` are doing the same things. Plus, uncovered lines make me sad. (_You_ did this to me!)

It's possible that I'm being overly paranoid, but heh. Your job is being a test-writing wizard, and my job is fretting about backward compatibility to ridiculous extremes. 😁

Hmm, there _are_ a few places where match groups other than 0 & 1 get checked for action & nick commands, but they're asserted to be `None`. We don't have _positive_ tests for groups in those command types to check that they get set correctly, only tests that verify they _aren't_ set when not expected.

The `rules` test stuff is enough to be a "separate PR" kind of thing now, so I'll leave it alone. But I'll reserve the right to take a stab at turning the deprecated functions into wrappers around the new stuff.

Comment (dgw) [reply test/test_tools.py]: Update: Pretty simple to turn the pattern functions into wrappers, actually. I should do the same for the regexp ones.

**Comment by dgw** [comment]
@Exirel Pushed a couple commits to address my own feedback about test coverage and stuff. Instead of -0.1% coverage, the last PR build reported only -0.004%. 🏆

When it's time to straighten out the PR history, I suggest:
* "fixup" the appropriate commit with the partial revert (currently 3176636)
* either leave e4cd829 standalone or merge it with 1f4df30 (along with the other amendment, 452818c)

**Comment by Exirel** [comment]
I'm really not fan of your solution here: when something is deprecated, it's for a reason. This should not be modified, and there is only one way to ensure that: don't change the code. Now, if we want to change the code in, say, the regex we use for good, we'll have to take care of *deprecated functions we don't want to change*.

**Comment by dgw** [comment]
These _deprecated functions we don't want to change_ are meant to return the regex or pattern for the type of command they specify. The regex isn't going to change between now and 8.0, at which point these will be gone.

But let's say we _do_ want to change something. We'd have made that change in _these_ functions before; now we make it in `plugins.rules` instead. I don't see a difference. If someone's code is going to break because the pattern/regex returned by this tool function changed, it's going to break regardless of whether said pattern/regex came from `tools/__init__.py` or `plugins/rules.py`. 🤷‍♂️

**Comment by Exirel** [comment]
@dgw rebased, keeping your changes. Should be good now.

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:1331]: Reusing the name `pattern` like this is a little bit weird, but I won't block on it.

**Review by dgw — APPROVED** [review]
Extra tests for the regex args, too, it's magnificent!

File: sopel/plugins/rules.py
Comment (Exirel) [reply sopel/plugins/rules.py:1331]: I was tempted to rename that as ` base_pattern` or ` command_pattern`, but heh, I'm fine with just `pattern`.

**Comment by Exirel** [comment]
Rebase because I want to work on the next step asap.

**Pull request merged**
